### ISSUE TITLE
Support for YAML templates

### ISSUE THREAD

**Issue by bgagnon** [logstash-plugins/logstash-output-elasticsearch#214]
YAML files are useful compared to JSON files because they are easier to read, maintain and support comments.

The template reading code at https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/master/lib/logstash/outputs/elasticsearch.rb#L479 assumes the file is JSON formatted.  

The template is parsed as JSON, then re-serialized later to JSON here:
https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/master/lib/logstash/outputs/elasticsearch/protocol.rb#L274

Supporting YAML files should only affect the first part. Arguably, if the template is never inspected by Ruby code, it could be sent verbatim to Elasticsearch.

**Comment by andrewvc** [comment]
@bgagnon thanks for the suggestion, but I think this adds an extra level of complexity to the plugin that isn't appropriate. This most likely belongs in a configuration management solution like Ansible or Chef. @jordansissel what are your thoughts?

**Comment by magnusbaeck** [comment]
I've written a tool that supports writing index templates in YAML and generating JSON from it. Apart from that conversion (which there obviously already are tools for) it allows you to attach short documentation strings to each field definition. Together with a list of which fields apply to which types one can easily generate field documentation for end users and filter authors. I can probably open source it if there's interest.

**Comment by jordansissel** [comment]
This yaml stuff sounds more like a feature for elasticsearch?

**Comment by magnusbaeck** [comment]
Yes, I think ES should be capable of getting YAML instead of JSON (for all REST calls, really).

**Comment by bgagnon** [comment]
Correct. Elasticsearch already supports YAML in its APIs. This issue is specific to the logstash plugin which prevents me from using YAML files for no apparent reason. Removing the parsing from the plugin simplifies it, in my opinion. It is Elasticsearch's responsibility to understand and make sense of the index template contents.

**Comment by andrewvc** [comment]
This is a good idea, but very low priority IMHO. I would gladly accept a patch however.
